gce_workload_certs_refresh/main.go

```diff
@@ -0,0 +1,290 @@
+	  }
+	 }
+	}
+*/
+
+// WorkloadIdentities represents Workload Identities in metadata.
+type WorkloadIdentities struct {
+	Status              string
+	WorkloadCredentials map[string]WorkloadCredential
+}
+
+// UnmarshalJSON is a custom JSON unmarshaller for WorkloadIdentities.
+func (wi *WorkloadIdentities) UnmarshalJSON(b []byte) error {
+	tmp := map[string]json.RawMessage{}
+	err := json.Unmarshal(b, &tmp)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(tmp["status"], &wi.Status); err != nil {
+		return err
+	}
+
+	wi.WorkloadCredentials = map[string]WorkloadCredential{}
+	wcs := map[string]json.RawMessage{}
+	if err := json.Unmarshal(tmp["workloadCredentials"], &wcs); err != nil {
+		return err
+	}
+
+	for domain, value := range wcs {
+		wc := WorkloadCredential{}
+		err := json.Unmarshal(value, &wc)
+		if err != nil {
+			return err
+		}
+		wi.WorkloadCredentials[domain] = wc
+	}
+
+	return nil
+}
+
+// WorkloadCredential represents Workload Credentials in metadata.
+type WorkloadCredential struct {
+	Metadata       Metadata
+	CertificatePem string
+	PrivateKeyPem  string
+}
+
+/*
+metadata key instance/workload-trusted-root-certs
+
+	{
+	 "status": "OK",
+	 "rootCertificates": {
+	  "PROJECT.svc.id.goog": {
+	   "metadata": {
+	    "workload_creds_dir_path": "/var/run/secrets/workload-spiffe-credentials"
+	   },
+	   "rootCertificatesPem": "-----BEGIN CERTIFICATE-----datahere-----END CERTIFICATE-----"
+	  }
+	 }
+	}
+*/
+
+// WorkloadTrustedRootCerts represents Workload Trusted Root Certs in metadata.
+type WorkloadTrustedRootCerts struct {
+	Status           string
+	RootCertificates map[string]RootCertificate
+}
+
+// UnmarshalJSON is a custom JSON unmarshaller for WorkloadTrustedRootCerts
+func (wtrc *WorkloadTrustedRootCerts) UnmarshalJSON(b []byte) error {
+	tmp := map[string]json.RawMessage{}
+	err := json.Unmarshal(b, &tmp)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(tmp["status"], &wtrc.Status); err != nil {
+		return err
+	}
+
+	wtrc.RootCertificates = map[string]RootCertificate{}
+	rcs := map[string]json.RawMessage{}
+	if err := json.Unmarshal(tmp["rootCertificates"], &rcs); err != nil {
+		return err
+	}
+
+	for domain, value := range rcs {
+		rc := RootCertificate{}
+		err := json.Unmarshal(value, &rc)
+		if err != nil {
+			return err
+		}
+		wtrc.RootCertificates[domain] = rc
+	}
+
+	return nil
+}
+
+// RootCertificate represents a Root Certificate in metadata
+type RootCertificate struct {
+	Metadata            Metadata
+	RootCertificatesPem string
+}
+
+// Metadata represents Metadata in metadata
+type Metadata struct {
+	WorkloadCredsDirPath string
+}
+
+func main() {
+	ctx := context.Background()
+
+	opts := logger.LogOpts{
+		LoggerName:     programName,
+		FormatFunction: logFormat,
+	}
+
+	opts.Writers = []io.Writer{os.Stderr}
+	logger.Init(ctx, opts)
+
+	// TODO: prune old dirs
+
+	if err := refreshCreds(); err != nil {
+		logger.Errorf(err.Error())
```

Comment:
isn't logger.Fatalf() more appropriate here?

Reply:
given it is at end of function, they're more or less interchangeable, but i wanted the 'done' message regardless of status.

Reply:
I understand but the process' exit code is not reflecting what really happened. Fatal will exit 1, even if you don't want to use fatal I'd recommend exiting a non 0 code.

Reply:
i think we want a 0 exit code because systemd can do different things based on failures. but i will double check this assumption and see if it needs to be updated.

Reply:
updated to use log.Fatal :)